core/primitives-core/src/parameter.rs

```diff
@@ -0,0 +1,202 @@
+    WasmStorageWriteEvictedByte,
+    WasmStorageReadBase,
+    WasmStorageReadKeyByte,
+    WasmStorageReadValueByte,
+    WasmStorageRemoveBase,
+    WasmStorageRemoveKeyByte,
+    WasmStorageRemoveRetValueByte,
+    WasmStorageHasKeyBase,
+    WasmStorageHasKeyByte,
+    WasmStorageIterCreatePrefixBase,
+    WasmStorageIterCreatePrefixByte,
+    WasmStorageIterCreateRangeBase,
+    WasmStorageIterCreateFromByte,
+    WasmStorageIterCreateToByte,
+    WasmStorageIterNextBase,
+    WasmStorageIterNextKeyByte,
+    WasmStorageIterNextValueByte,
+    WasmTouchingTrieNode,
+    WasmReadCachedTrieNode,
+    WasmPromiseAndBase,
+    WasmPromiseAndPerPromise,
+    WasmPromiseReturn,
+    WasmValidatorStakeBase,
+    WasmValidatorTotalStakeBase,
+    WasmAltBn128G1MultiexpBase,
+    WasmAltBn128G1MultiexpElement,
+    WasmAltBn128PairingCheckBase,
+    WasmAltBn128PairingCheckElement,
+    WasmAltBn128G1SumBase,
+    WasmAltBn128G1SumElement,
+
+    // Smart contract limits
+    MaxGasBurnt,
+    MaxGasBurntView,
+    MaxStackHeight,
+    StackLimiterVersion,
+    InitialMemoryPages,
+    MaxMemoryPages,
+    RegistersMemoryLimit,
+    MaxRegisterSize,
+    MaxNumberRegisters,
+    MaxNumberLogs,
+    MaxTotalLogLength,
+    MaxTotalPrepaidGas,
+    MaxActionsPerReceipt,
+    MaxNumberBytesMethodNames,
+    MaxLengthMethodName,
+    MaxArgumentsLength,
+    MaxLengthReturnedData,
+    MaxContractSize,
+    MaxTransactionSize,
+    MaxLengthStorageKey,
+    MaxLengthStorageValue,
+    MaxPromisesPerFunctionCallAction,
+    MaxNumberInputDataDependencies,
+    MaxFunctionsNumberPerContract,
+    Wasmer2StackLimit,
+    MaxLocalsPerContract,
+}
+
+pub fn load_parameters_from_txt(arg: &[u8]) -> BTreeMap<Parameter, String> {
+    let parameters = BufReader::new(arg).lines().filter_map(|line| {
+        let trimmed = line.unwrap().trim().to_owned();
+        if trimmed.starts_with("#") || trimmed.is_empty() {
+            None
+        } else {
+            let mut iter = trimmed.split(":");
+            let key_str = iter.next().unwrap().trim();
+            let typed_key = Parameter::from_str(key_str)
+                .unwrap_or_else(|_err| panic!("Unexpected parameter `{key_str}`."));
+            let value =
+                iter.next().expect("Parameter name and value must be separated by a ':'").trim();
+            Some((typed_key.to_owned(), value.to_owned()))
+        }
+    });
+    BTreeMap::from_iter(parameters)
+}
+
+pub fn read_parameter<F: std::str::FromStr>(
+    params: &BTreeMap<Parameter, String>,
+    key: Parameter,
+) -> F {
+    let key_str: &'static str = key.into();
+    read_optional_parameter(params, key).unwrap_or_else(|| panic!("Missing parameter `{key_str}`"))
+}
+
+pub fn read_optional_parameter<F: std::str::FromStr>(
+    params: &BTreeMap<Parameter, String>,
+    key: Parameter,
+) -> Option<F> {
+    let key_str: &'static str = key.into();
+    Some(params.get(&key)?.parse().unwrap_or_else(|_err| {
```

Comment:
Hm, I'm not sure if `FromStr` is the best tool to use here. Notably there is no implementation of `FromStr` for things like `Option<_>` and so there is no way to reset a parameter to `None` if the parameter was already set to `Some` before.

Reply:
I think I now got rid of all uses of `FromStr`. Parsing to types is mostly left to `serde`. Although, the semantics of empty, missing, or null values is still a bit awkward